File: 632.smallest-range-covering-elements-from-k-lists.cs
/*
 * @lc app=leetcode id=632 lang=csharp
 *
 * [632] Smallest Range Covering Elements from K Lists
 */

// @lc code=start
using System;
using System.Collections.Generic;
public class Solution 
{
    public int[] SmallestRange(IList<IList<int>> nums) 
    {
        int globalMax = 100000;
        int globalMin = - 100000;
        int maxSoFar = - 100000;  
        PriorityQueue<int[], int> pq = new PriorityQueue<int[], int>();

        for (int i = 0; i <  nums.Count; i++)
        {
            globalMax = Math.Max(globalMax, nums[i][nums[i].Count - 1]);
            globalMin = Math.Min(globalMin, nums[i][0]);
            pq.Enqueue(new int[2]{i, 0}, nums[i][0]);
            maxSoFar = Math.Max(maxSoFar, nums[i][0]);
        }   

        int[] result = new int[2]{globalMin, globalMax};
        int diff = globalMax - globalMin;

        while(pq.TryDequeue(out int[] node, out int num))
        {
            if (maxSoFar - num < diff)
            {
                diff = maxSoFar - num;
                result[0] = num;
                result[1] = maxSoFar;
            }
            if (node[1] < nums[node[0]].Count - 1)
            {
                pq.Enqueue(new int[2]{node[0], node[1] + 1}, nums[node[0]][node[1] + 1]);
                maxSoFar = Math.Max(maxSoFar, nums[node[0]][node[1] + 1]);
            }
            else
            {
                break;
            }
        }
        return result;
    }
}
// @lc code=end

File: 632.smallest-range-covering-elements-from-k-lists.py
#
# @lc app=leetcode id=632 lang=python3
#
# [632] Smallest Range Covering Elements from K Lists
#

# @lc code=start
import heapq
from typing import List


class Solution:
    '''        
    The idea is to go through one of the lists from min to max
    at the beginning, push the first element of each list with its row and column indices into a min heap. Then at each round, find min element from the heap. Get the next larger number in that row. We keep track of the largest element inside the heap. Because in each pop, we always remove one element and add back another element to the heap, the heap always keeps one element of each list inside the heap. The difference between the largest element in the heap (maxsofar) and the smallest element (the front of the heap) is kept tracked of. The process ends when one of the lists arrived at the end (further element addition from a different row will violate the one row - one element condition). 
    '''
    def smallestRange(self, nums: List[List[int]]) -> List[int]:
        globalMax, globalMin, maxSoFar = - float('inf'), float('inf'), - float('inf')
        heap = []
        for i, num in enumerate(nums):
            globalMax = max(globalMax, num[-1])
            globalMin = min(globalMin, num[0])
            heapq.heappush(heap, [num[0], i, 0])
            maxSoFar = max(maxSoFar,num[0])
            
        result = [globalMin, globalMax]
        diff = globalMax - globalMin
        
        while heap:
            num, row, col = heapq.heappop(heap)
            if maxSoFar - num < diff:
                diff = maxSoFar - num
                result = [num, maxSoFar]
            if col < len(nums[row]) - 1:
                heapq.heappush(heap, [nums[row][col + 1], row, col + 1])
                maxSoFar = max(maxSoFar, nums[row][col + 1])
            else:
                break
        return result
# @lc code=end